Sources/DeepLearning/Operators/Basic.swift

```diff
@@ -0,0 +1,303 @@
+    /// 
+    /// // t3 has shape [2, 3]
+    /// // t4 has shape [2, 3]
+    /// t3.concatenated(with: [t4]) // has shape [4, 3]
+    /// t3.concatenated(with: [t4], alongAxis: 1) // has shape [2, 6]
+    /// ```
+    ///
+    /// - Note: If you are concatenating along a new axis consider using `stacked`.
+    ///
+    /// - Parameters:
+    ///   - tensors: Tensors to concatenate with the current tensor.
+    ///   - axis: Dimension along which to concatenate. Negative values wrap around.
+    ///
+    /// - Precondition: All tensors must have the same rank as the current tensor and all dimensions 
+    ///     except `axis` must be equal.
+    /// - Precondition: `axis` must be in the range `[-rank, rank)`.
+    /// 
+    /// - Returns: The concatenated tensor.
+    @inlinable
+    // @differentiable(vjp: _vjpConcatenated where Scalar : TensorFlowFloatingPoint)
+    func concatenated(with tensors: [Tensor], alongAxis axis: Int32 = 0) -> Tensor {
+        return Raw.concatV2([self] + tensors, axis: Tensor<Int32>(axis))
+    }
+
+    /// Gathers slices of this tensor at `indices` along the `axis` dimension.
+    ///
+    /// For 0-D (scalar) `indices`:
+    /// ```
+    /// result[p_0,          ..., p_{axis-1},
+    ///        p_{axis + 1}, ..., p_{N-1}] = 
+    /// self[p_0,          ..., p_{axis-1},
+    ///      indices,
+    ///      p_{axis + 1}, ..., p_{N-1}]
+    /// ```
+    /// 
+    /// For 1-D (vector) `indices`:
+    /// ```
+    /// result[p_0,          ..., p_{axis-1},
+    ///        i,
+    ///        p_{axis + 1}, ..., p_{N-1}] = 
+    /// self[p_0,          ..., p_{axis-1},
+    ///      indices[i],
+    ///      p_{axis + 1}, ..., p_{N-1}]
+    /// ```
+    /// 
+    /// In the general case, produces a resulting tensor where:
+    /// ```
+    /// result[p_0,             ..., p_{axis-1},
+    ///        i_{batch\_dims}, ..., i_{M-1},
+    ///        p_{axis + 1},    ..., p_{N-1}] = 
+    /// self[p_0,             ..., p_{axis-1},
+    ///      indices[i_0,     ..., i_{M-1}],
+    ///      p_{axis + 1},    ..., p_{N-1}]
+    /// ```
+    /// where `N = self.rank` and `M = indices.rank`.
+    ///
+    /// The shape of the resulting tensor is:
+    /// `self.shape[..<axis] + indices.shape + self.shape[(axis + 1)...]`.
+    /// 
+    /// - Note: On CPU, if an out-of-range index is found, an error is thrown. On GPU, if an 
+    /// out-of-range index is found, a 0 is stored in the corresponding output values.
+    ///
+    /// - Parameters:
+    ///   - indices: Contains the indices to gather.
+    ///   - axis: Dimension along which to gather. Negative values wrap around.
+    /// 
+    /// - Precondition: `axis` must be in the range `[-rank, rank)`.
+    /// 
+    /// - Returns: The gathered tensor.
+    @inlinable
+    // @differentiable(vjp: _vjpGathering where Scalar: TensorFlowFloatingPoint)
+    func gathering<I: TensorFlowInteger>(
+        atIndices indices: Tensor<I>, 
```

Comment:
Since our index type is `Int32`, this argument should be `Tensor<Int32>` for consistency, instead of being generic. Same for other APIs in this file that take indices.